src/code/sys_flashrom.c

```diff
@@ -3,42 +3,242 @@
+        }
+    }
+    return -1;
+}
+
+s32 SysFlashrom_InitFlash(void) {
+    osCreateMesgQueue(&sFlashromMesgQueue, sFlashromMesg, ARRAY_COUNT(sFlashromMesg));
+    osFlashInit();
+    sFlashromFlashIsGood = true;
+    if (SysFlashrom_CheckFlashType()) {
+        sFlashromFlashIsGood = false;
+        return -1;
+    }
+    return 0;
+}
+
+s32 SysFlashrom_ReadData(void* addr, s32 pageNum, s32 pageCount) {
+    OSIoMesg msg;
+
+    if (!SysFlashrom_FlashIsGood()) {
+        return -1;
+    }
+    osInvalDCache(addr, pageCount * 128);
+    osFlashReadArray(&msg, 0, pageNum, addr, pageCount, &sFlashromMesgQueue);
+    osRecvMesg(&sFlashromMesgQueue, NULL, OS_MESG_BLOCK);
+    return 0;
+}
+
+s32 SysFlashrom_ErasePage(u32 page) {
+    if (!SysFlashrom_FlashIsGood()) {
+        return -1;
+    }
+    return osFlashSectorErase(page);
+}
+
+s32 SysFlashrom_WriteData(u8* addr, u32 pageNum, u32 pageCount) {
+    OSIoMesg msg;
+    s32 result;
+    u32 i;
+
+    if (!SysFlashrom_FlashIsGood()) {
+        return -1;
+    }
+    if (pageNum & 0x7F) {
+        Fault_AddHungupAndCrash("../sys_flashrom.c", 275);
+    }
+    osWritebackDCache(addr, pageCount * 128);
+    for (i = 0; i != pageCount; i++) {
+        osFlashWriteBuffer(&msg, 0, addr + i * 128, &sFlashromMesgQueue);
+        osRecvMesg(&sFlashromMesgQueue, NULL, OS_MESG_BLOCK);
+        result = osFlashWriteArray(i + pageNum);
+        if (result != 0) {
+            return result;
+        }
+    }
+    return 0;
+}
+
+s32 SysFlashrom_AttemptWrite(void* addr, s32 pageNum, s32 pageCount) {
+    s32 result;
+    s32 i;
+
+    if (!SysFlashrom_FlashIsGood()) {
+        return -1;
+    }
+    osWritebackDCache(addr, pageCount * 128);
+    i = 0;
+failRetry:
+    result = SysFlashrom_ErasePage(pageNum);
+    if (result != 0) {
+        if (i < 3) {
+            i++;
+            goto failRetry;
+        }
+        return result;
+    }
+    result = SysFlashrom_WriteData(addr, pageNum, pageCount);
+    if (result != 0) {
+        if (i < 3) {
+            i++;
+            goto failRetry;
+        }
+        return result;
+    }
+    return 0;
+}
+
+s32 func_80185BE4(void* data, void* addr, u32 pageCount) {
+    u32 size;
+    u32 i;
+
+    for (i = 0; i < pageCount * 128; i += 4) {
+        if ((*(s32*)data & *(s32*)addr) != *(s32*)addr) {
+            return 0;
+        }
+    }
+    return 1;
+}
+
+s32 SysFlashrom_WriteData2(void* addr, u32 pageNum, u32 pageCount) {
+    void* data;
+    size_t size;
+    s32 ret;
+
+    if (!SysFlashrom_FlashIsGood()) {
+        return -1;
+    }
+    size = pageCount * 128;
+    data = SystemArena_Malloc(size);
+    if (data == 0) {
+        ret = SysFlashrom_AttemptWrite(addr, pageNum, pageCount);
+    } else {
+        SysFlashrom_ReadData(data, pageNum, pageCount);
+        if (bcmp(data, addr, size) == 0) {
+            ret = 0;
+        } else {
+            //! @bug Will attempt to write regardless of the result of the function call.
```

Comment:
Given the prior explanation of what `func_80185BE4` tries to do, it's not really a bug to run a write either way, I would say if anything the real bug is that it runs an erasure either way. I'm not entirely sure given how it doesn't really cause any bad side-effects whether it's worthy of "bug", but it is worth noting in some way for sure.